### ISSUE TITLE
deprecation of `int`

### ISSUE THREAD

**Issue by my-little-repository** [JuliaLang/julia#13126]
Can we please undeprecate the `int` function?

In my opinion, it is as useful as the `big` and `float` functions. I use it e.g. when a type is lost in a computation, which happens when methods/algorithms are not yet implemented for integer arguments.

    julia> det([-5 -5 ; -3 -5])
    9.999999999999998
    julia> Int(det([-5 -5 ; -3 -5]))
    ERROR: InexactError()
     in call at essentials.jl:56
    julia> int(det([-5 -5 ; -3 -5]))
    10

Note also that Integers are used as indices to arrays. With the removal of Reals indices, the `int` function will become even more useful. Here is an example where I think that `int` is the better choice with respect to readability.

     A[int((m+1)/2)]
     A[convert(Int, (m+1)/2)]
     A[(m+1) >> 1]

Here are a few constructs that are proposed by the Warnings in order to palliate the deprecation of `int`.

    map(Int, x)
    round(Int, x)
    parse(Int, x)
    Complex((Int64)(real(z)),(Int64)(imag(z)))

The fact that there is at least four different nonobvious constructs needed to replace `int` should point out to the need for such a function.

**Comment by JeffBezanson** [comment]
> four different nonobvious constructs needed to replace int should point out to the need for such a function

Quite the opposite --- I can't comment on obviousness, but it's quite important to specify whether a certain case needs round, floor, ceil, or trunc, and `int(x)` obscures that distinction. Combining parsing into the same function is even stranger.

You can use `Int(x)` instead of `convert(Int, x)`.

**Comment by pkofod** [comment]
^ +1

I agree that if we can expect people to think int() does four (significantly) different things, it should not exist under that name.

**Comment by JeffreySarnoff** [comment]
+1 let go of "int"

**Comment by tbreloff** [comment]
+1... I like terse code, but it should go.  While we're at it, though...
lets deprecate `float(x)` in the same vein, since you're still likely
rounding when you create a floating point value.  the only non-deprecated
lower-case number method allowed??  `unum(x)`  :)

**Comment by my-little-repository** [comment]
@pfokod In that piece of code, m was known to be odd. Due to the deprecation of Reals indices, there was the need to explain to the compiler that the index was in fact an integer. Just like in the first piece of code, `int` does the job neatly. I ran several times into problems when I tried  using `Int` in place of `int`. For example, in the computation of the determinant of an integer matrix, Int works for some matrices but not all. I think that `int` has really the DWIM flavor, a testimony to the work that went into its implementation.

So basically the counterargument is "polymorphic behavior is evil". If this leads to the deprecation of `float`, that will be the most counterproductive bug report I ever made!

**Comment by nalimilan** [comment]
@my-little-repository Why don't you use `round` instead of `int`? That's what it will do anyway if the value cannot be converted exactly to an integer, and I would consider it as the true DWIM solution.

**Comment by my-little-repository** [comment]
@nalimilan `round` does not work on arrays or strings and it is more verbose because you have to specify the return type.

I am not saying that `int` is  irreplaceable, just that it is quite handy and does not harm anybody. I can live without it being part of the standard library though. Since nobody spoke in favor of its return, I think it's best to close the issue.

**Comment by nalimilan** [comment]
It does hurt in the sense that people get confused between `int` and `Int`, and if they use the former they lose the protection from undesired rounding.

One alternative is to make it easier to apply *any* function to an array (https://github.com/JuliaLang/julia/issues/8450).

**Comment by JeffBezanson** [comment]
> So basically the counterargument is "polymorphic behavior is evil".

No, it's that "DWIM" is evil, because a computer doesn't know what you *mean* --- nor does anybody reading the code, nor (sometimes in my experience) even myself.

**Comment by ScottPJones** [comment]
To me, either `A[div(m+1,2)]` or `A[(m+1)>>1]` make more sense than the deprecated `A[int((m+1)/2)]`. Of course, I would have preferred it if `/` acted as `div` when both arguments are integers.  Elsewhere, Julia tries to maintain some type stability, but `/` when all arguments are integers -> `Float` just seems inconsistent (and wrong).
Also, `int((m+1)/2)` or `convert(Int, (m+1)/2)` won't even get you what you expect, if m has more than 53 bits!

**Comment by tkelman** [comment]
Look up the pep on integer division please. The 53 bits precision issue could be addressed in the future if doable without much performance cost.

**Comment by quinnj** [comment]
It's a tough call @ScottPJones and if I remember correctly, there have been several long and hard discussions about `Int / Int => Float`. The balance is between getting a "safe" answer that's probably what you expect (particularly for those more familiar with python, matlab, R, etc.) and the more C-like approach. I imagine it rubs you wrong since you're more of a low-level, C-type dev.

**Comment by ScottPJones** [comment]
No, it rubs me wrong because it is silent data loss, has nothing to do with C.
The implicit conversion wouldn't bother me *if* it used a "safe_convert" function, that would give an `InexactError` if the integer could not be converted to a float without loss.

**Comment by carlobaldassi** [comment]
> ```julia
A[int((m+1)/2)]
A[convert(Int, (m+1)/2)]
A[(m+1) >> 1]
```

Arguably, using `\div` produces the most readable and compact form of all:

```julia
A[(m+1)÷2]
```

**Comment by ScottPJones** [comment]
More importantly, `(m+1)÷2` always gives the correct result, unlike any that have `(m+1)/2`.

**Comment by ScottPJones** [comment]
Here's some more fun:
```julia
julia> ((m+1)÷65535)
0x0000800080008000

julia> convert(Int, (m+1)/65535)
ERROR: InexactError()
 in convert at int.jl:196

julia> (m+1)/65535
1.407396358717445e14

julia> m
0x7ffffffffffffffe
```
